3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,342 @@
+  )
+}
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a scalar or a JSON field) along with
+ * a function to apply that strategy to the EnrichedEvent POJO (A scalar field is represented in config py "pojo")
+ */
+sealed trait PiiField {
+  import PiiConstants.Mutator
+
+  /**
+   * Strategy for this field
+   *
+   * @return PiiStrategy a strategy to be applied to this field
+   */
+  def strategy: PiiStrategy
+
+  /**
+   * The POJO mutator for this field
+   *
+   * @return fieldMutator
+   */
+  def fieldMutator: Mutator
+
+  /**
+   * Gets an enriched event from the enrichment manager and modifies it according to the specified strategy.
+   *
+   * @param event The enriched event
+   */
+  def transform(event: EnrichedEvent): Unit = fieldMutator(event, applyStrategy)
+
+  protected def applyStrategy(fieldValue: String): String
+}
+
+/**
+ * PiiStrategy trait. This corresponds to a strategy to apply to a single field. Currently only String input is
+ * supported.
+ */
+sealed trait PiiStrategy {
+  def scramble(clearText: String): String
+}
+
+/**
+ * Companion object. Lets us create a PiiPseudonymizerEnrichment
+ * from a JValue.
+ */
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+  import PiiConstants._
+
+  implicit val json4sFormats = DefaultFormats
+
+  override val supportedSchema =
+    SchemaCriterion("com.snowplowanalytics.snowplow.enrichments", "pii_enrichment_config", "jsonscehma", 1, 0, 0)
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    for {
+      conf <- matchesSchema(config, schemaKey)
+      enabled = ScalazJson4sUtils.extract[Boolean](conf, "enabled").toOption.getOrElse(false)
+      piiFields        <- ScalazJson4sUtils.extract[List[JObject]](conf, "parameters", "pii").leftMap(_.getMessage)
+      strategyFunction <- extractStrategyFunction(config)
+      hashFunction     <- getHashFunction(strategyFunction)
+      piiFieldList     <- extractFields(piiFields, PiiStrategyPseudonymize(hashFunction))
+    } yield if (enabled) PiiPseudonymizerEnrichment(piiFieldList) else PiiPseudonymizerEnrichment(List())
+  }.leftMap(_.toProcessingMessageNel)
+
+  private def getHashFunction(strategyFunction: String): Validation[String, MessageDigest] =
+    try {
+      MessageDigest.getInstance(strategyFunction).success
+    } catch {
+      case e: NoSuchAlgorithmException =>
+        s"Could not parse PII enrichment config: ${e.getMessage()}".failure
+    }
+
+  private def extractFields(piiFields: List[JObject], strategy: PiiStrategy): Validation[String, List[PiiField]] =
+    piiFields.map {
+      case JObject(List(("pojo", JObject(List(("field", JString(fieldName))))))) =>
+        ScalarMutators
+          .get(fieldName)
+          .map(PiiScalar(strategy, _).success)
+          .getOrElse(s"The specified pojo field ${fieldName} is not supported".failure)
+      case JObject(List(("json", jsonField))) =>
```

Comment:
Feel free to ignore, but I'd like to not make a precedent of pattern-matching JSONs this way. It's quite dangerous if for example junior developer who doesn't know that `keys` in `JObject(keys)` is order-sensetive.

Reply:
@chuwy It's easy for anyone to miss that especially if you haven't used json4s. In this case there is a single value and this is very convenient, but if there is a better, clearer way to get that value out I would like to know.